pkg/timeutils/utils.go

```diff
@@ -1,36 +1,122 @@
+
 	if strings.Contains(value, ".") {
-		format = time.RFC3339Nano
+		if parseInLocation {
+			format = RFC3339NanoLocal
+		} else {
+			format = time.RFC3339Nano
+		}
+	} else if strings.Contains(value, "T") {
+		// we want the number of colons in the T portion of the timestamp
+		tcolons := strings.Count(value, ":")
+		// if parseInLocation is off and we have a +/- zone offset (not Z) then
+		// there will be an extra colon in the input for the tz offset subract that
+		// colon from the tcolons count
+		if !parseInLocation && !strings.ContainsAny(value, "zZ") && tcolons > 0 {
+			tcolons--
+		}
+		if parseInLocation {
+			if tcolons >= 2 {
+				format = RFC3339Local
+			} else if tcolons == 1 {
+				format = "2006-01-02T15:04"
+			} else {
+				format = "2006-01-02T15"
+			}
+		} else {
+			if tcolons >= 2 {
+				format = time.RFC3339
+			} else if tcolons == 1 {
+				format = "2006-01-02T15:04Z07:00"
+			} else {
+				format = "2006-01-02T15Z07:00"
+			}
+		}
+	} else if parseInLocation {
+		format = DateLocal
+	} else {
+		format = DateWithZone
+	}
+
+	var t time.Time
+	var err error
+
+	if parseInLocation {
+		t, err = time.ParseInLocation(format, value, time.FixedZone(time.Now().Zone()))
 	} else {
-		format = time.RFC3339
+		t, err = time.Parse(format, value)
 	}
 
-	loc := time.FixedZone(time.Now().Zone())
-	if len(value) < len(format) {
-		format = format[:len(value)]
+	if err != nil {
+		// if there is a `-` then its an RFC3339 like timestamp otherwise assume unixtimestamp
+		if strings.Contains(value, "-") {
+			return "", err // was probably an RFC3339 like timestamp but the parser failed with an error
+		}
+		return value, nil // unixtimestamp in and out case
+	}
+
+	return fmt.Sprintf("%d.%09d", t.Unix(), int64(t.Nanosecond())), nil
+}
+
+// ParseTimestamps returns seconds and nanoseconds from a timestamp that has the
+// format "%d.%09d", time.Unix(), int64(time.Nanosecond()))
+// if the incoming nanosecond portion is longer or shorter than 9 digits it is
+// converted to nanoseconds.  The expectation is that the seconds and
+// seconds will be used to create a time variable.  For example:
+//     seconds, nanoseconds, err := ParseTimestamp("1136073600.000000001",0)
+//     if err == nil since := time.Unix(seconds, nanoseconds)
+// returns seconds as def(aultSeconds) if value == ""
+func ParseTimestamps(value string, def int64) (int64, int64, error) {
+	if value == "" {
+		return def, 0, nil
+	}
+	sa := strings.SplitN(value, ".", 2)
+	s, err := strconv.ParseInt(sa[0], 10, 64)
+	if err != nil {
+		return s, 0, err
+	}
+	if len(sa) != 2 {
+		return s, 0, nil
 	}
-	t, err := time.ParseInLocation(format, value, loc)
+	n, err := strconv.ParseInt(sa[1], 10, 64)
 	if err != nil {
-		return value
+		return s, n, err
 	}
-	return strconv.FormatInt(t.Unix(), 10)
+	// should already be in nanoseconds but just in case convert n to nanoseonds
+	n *= int64(math.Pow(float64(10), float64(9-len(sa[1]))))
```

Comment:
Note the unit testcases that were added:
		{"1136073600.000000001", 0, 1136073600, 1},
		{"1136073600.0000000010", 0, 1136073600, 1},
		{"1136073600.00000001", 0, 1136073600, 10},

The target output is in nanoseconds because that's the variable needed for the unixnano time function, if the length is > 9 the code above shifts the decimal place to the left, if the length is less than 9 it shifts the decimal place to the right (mathematically).  Could've done this with string functions instead of math by truncating (substr) it to 9 digits if len(sa[1])>9 and adding the right number of zeros to the end to make it 9 in length if len(sa[1])< 9 before converting sa[1] to an int64...